### PR TITLE
Provide cached visibility + retries of BPF maps via API

### PR DESCRIPTION
**Pull request by tgraf**
This is a re-spin of #4792 after #4843 has been addressed.

This PR started out due to lack of visibility of LPM based bpf maps which cannot be dumped. The first commit introduces a new cache per bpf map which can be enabled with WithCache() that is then kept up to date based on Update() and Delete() functions called. The functionality is enabled on all user space controller maps except for the policymap which requires rework to use bpf.Map.

The second commit makes use of the cache and implements automatic retries on errors. It is primarily useful for BPF maps which are fed by events from the kvstore such as the ipcache or k8s such as services as there is no feedback mechanism.

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/4854)

### CODE REVIEW

**Comment by tgraf** [comment]
test-me-please

**Comment by aanm** [comment]
@tgraf I'm going to trigger a `test-missed-k8s` since they were the ones to detect the failure in the first place

**Comment by aanm** [comment]
test-missed-k8s

**Review by aanm — APPROVED** [review]

**Comment by tgraf** [comment]
Hit #4852

**Comment by tgraf** [comment]
test-missed-k8s

**Comment by tgraf** [comment]
test-me-please

**Comment by tgraf** [comment]
Hit #4852

**Comment by tgraf** [comment]
test-me-please

File: cilium/cmd/map_get.go [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+		if m == nil {
+			return
+		}
+
+		if command.OutputJSON() {
+			if err := command.PrintOutput(m); err != nil {
+				os.Exit(1)
+			}
+		} else {
+			w := tabwriter.NewWriter(os.Stdout, 5, 0, 3, ' ', 0)
```
Comment (ianvernon) [inline cilium/cmd/map_get.go]: I see that the functionality here in `map_list.go` L73-L82 is basically the same, would be nice to factorize it into a function.

File: pkg/bpf/map.go [hunk 2]
```diff
@@ -439,25 +496,73 @@ func (m *Map) Lookup(key MapKey) (MapValue, error) {
...
+	if err = m.Open(); err != nil {
 		return err
 	}
 
-	return UpdateElement(m.fd, key.GetKeyPtr(), value.GetValuePtr(), 0)
+	err = UpdateElement(m.fd, key.GetKeyPtr(), value.GetValuePtr(), 0)
+	return err
+}
+
+func (m *Map) deleteCacheEntry(key MapKey, err error) {
```
Comment (ianvernon) [inline pkg/bpf/map.go]: Add a comment that specifies that this assumes the map's mutex is being held.

**Review by ianvernon — APPROVED** [review]
LGTM, some small nits that can be addressed separately.

File: cilium/cmd/map_get.go (hunk 1 above)
Comment (tgraf) [reply cilium/cmd/map_get.go]: Fixed

File: pkg/bpf/map.go (hunk 2 above)
Comment (tgraf) [reply pkg/bpf/map.go]: Fixed

**Comment by tgraf** [comment]
test-me-please

**Pull request merged**